### PR TITLE
RESP3 tests

### PR DESCRIPTION
**Pull request by dvora-h**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [ ] Does `$ tox` pass with this change (including linting)?
- [ ] Do the CI tests pass with this change (enable it first in your forked repo and wait for the github action build to finish)?
- [ ] Is the new or changed code fully tested?
- [ ] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?
- [ ] Is there an example added to the examples folder (if applicable)?
- [ ] Was the change added to CHANGES file?

_NOTE: these things are not required to open a PR and can be done
afterwards / while the PR is open._

### Description of change

_Please provide a description of the change here._

### CODE REVIEW

File: redis/asyncio/client.py [hunk 1]
```diff
@@ -672,12 +672,12 @@ def __init__(
             self.encoder = self.connection_pool.get_encoder()
         if self.encoder.decode_responses:
             self.health_check_response: Iterable[Union[str, bytes]] = [
```
Comment (chayim) [inline redis/asyncio/client.py:674]: IMHO you can remove the Iterable piece in 674. I know you didn't add it, but it's in function, so has no value.

File: redis/asyncio/cluster.py [hunk 2]
```diff
@@ -319,6 +319,8 @@ def __init__(
             kwargs.update({"retry": self.retry})
 
         kwargs["response_callbacks"] = self.__class__.RESPONSE_CALLBACKS.copy()
+        if kwargs.get("protocol") in ["3", 3]:
```
Comment (chayim) [inline redis/asyncio/cluster.py:322]: I like that this is a one time cost

File: redis/asyncio/connection.py [hunk 3]
```diff
@@ -333,16 +333,35 @@ def _error_message(self, exception):
...
             cred_provider = (
                 self.credential_provider
                 or UsernamePasswordCredentialProvider(self.username, self.password)
             )
             auth_args = cred_provider.get_credentials()
-            # avoid checking health here -- PING will fail if we try
-            # to check the health prior to the AUTH
+            # if resp version is specified and we have auth args,
+            # we need to send them via HELLO
+        if auth_args and self.protocol != 2:
```
Comment (chayim) [inline redis/asyncio/connection.py:348]: Same issue? self.protocol not in ['2', 2]?

File: redis/asyncio/connection.py [hunk 4]
```diff
@@ -333,16 +333,35 @@ def _error_message(self, exception):
...
+            if isinstance(self._parser, _AsyncRESP2Parser):
+                self.set_parser(_AsyncRESP3Parser)
+                # update cluster exception classes
+                self._parser.EXCEPTION_CLASSES = parser.EXCEPTION_CLASSES
+                self._parser.on_connect(self)
+            if len(auth_args) == 1:
+                auth_args = ["default", auth_args[0]]
+            await self.send_command("HELLO", self.protocol, "AUTH", *auth_args)
+            response = await self.read_response()
+            if response.get(b"proto") != int(self.protocol) and response.get(
```
Comment (chayim) [inline redis/asyncio/connection.py:358]: will have to be changed to accomodate that

File: redis/client.py [hunk 5]
```diff
@@ -331,9 +331,15 @@ def parse_xinfo_stream(response, **options):
             data["last-entry"] = (last[0], pairs_to_dict(last[1]))
     else:
         data["entries"] = {_id: pairs_to_dict(entry) for _id, entry in data["entries"]}
-        data["groups"] = [
-            pairs_to_dict(group, decode_keys=True) for group in data["groups"]
-        ]
+        if isinstance(data["groups"][0], list):
```
Comment (chayim) [inline redis/client.py:334]: try/except instead == cheaper

File: redis/cluster.py [hunk 6]
```diff
@@ -1633,7 +1642,10 @@ def __init__(self, redis_cluster, node=None, host=None, port=None, **kwargs):
         self.node_pubsub_mapping = {}
         self._pubsubs_generator = self._pubsubs_generator()
         super().__init__(
-            **kwargs, connection_pool=connection_pool, encoder=redis_cluster.encoder
+            **kwargs,
```
Comment (chayim) [inline redis/cluster.py:1645]: pass the kwargs at the end of the init...?

File: redis/cluster.py [hunk 7]
```diff
@@ -1717,14 +1729,18 @@ def execute_command(self, *args):
...
+                self.connection._parser.set_push_handler(self.push_handler_func)
         connection = self.connection
         self._execute(connection, connection.send_command, *args)
 
     def _get_node_pubsub(self, node):
         try:
             return self.node_pubsub_mapping[node.name]
         except KeyError:
-            pubsub = node.redis_connection.pubsub()
+            pubsub = node.redis_connection.pubsub(
```
Comment (chayim) [inline redis/cluster.py:1741]: good!

File: redis/client.py (hunk 5 above)
Comment (dvora-h) [reply redis/client.py:334]: It won't work here because of the content of data["groups"]

**Review by chayim — APPROVED** [review]

File: redis/asyncio/connection.py (hunk 3 above)
Comment (vladvildanov) [reply redis/asyncio/connection.py]: Isn't `int(self.protocol) != 2` more readable?

**Pull request merged**